File: oo-lib-py/app/v1/member.py
from .catalog import Catalog

class Member: 
    def __init__(self, name: str) -> None:
        self._name = name
        self._borrowed_ids = [] # list for borrowed Ids
        self._MAX_BORROW_LIMIT = 5

    @property
    def name(self) -> str:
        return self._name

    def borrow(self, id: str) -> None:
        # Check for max borrow limit
        if len(self._borrowed_ids) >= self._MAX_BORROW_LIMIT:
            raise ValueError("Maximum borrowed items reached!")

        # Check for duplicate borrows
        if id in self._borrowed_ids:
            raise ValueError(f"Duplicate borrow for id: {id}")
        self._borrowed_ids.append(id)

    def borrowed_ids(self) -> list:
        return self._borrowed_ids

    def return_item(self, id: str) -> None:
        self._borrowed_ids.remove(id)

    def list_details(self, cat: Catalog) -> list:
        lines = [] 

        for id in self._borrowed_ids:
            item = cat.get(id)
            lines.append(f"{item._title} ({item.days_allowed()} days)")

        return lines
    
    def is_borrowed_list_valid(self, cat: Catalog) -> bool:
        for id in self._borrowed_ids:
            if cat.get(id) == None:
                return False
        return True

File: oo-lib-rust/src/v1/member.rs
use crate::v1::catalog::Catalog;

pub struct Member {
    name: String,
    borrowed_ids: Vec<String>,
    max_borrow_limit: usize,
}

impl Member {
    pub fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            borrowed_ids: Vec::new(),
            max_borrow_limit: 5,
        }
    }
    
    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn borrowed_ids(&self) -> &Vec<String> {
        &self.borrowed_ids
    }

    pub fn borrow(&mut self, id: &str) -> Result<(), String> {
        // Check for max borrow limit
        if self.borrowed_ids.len() == self.max_borrow_limit {
            Err(format!("Max borrow limit reached!"))
        
        // Check for duplicate borrows
        } else if self.borrowed_ids.contains(&id.to_string()) {
            Err(format!("Duplicate borrow for id: {}", id))

        // Else add to list
        } else {
            self.borrowed_ids.push(id.to_string());
            Ok(())
        }
    }

    pub fn return_item(&mut self, id: &str) -> Result<(), String> {
        if self.borrowed_ids.contains(&id.to_string()) {
            self.borrowed_ids.retain(|x| *x != id);
            Ok(())
        } else {
            Err(format!("No item found with id: {}", id))
        }
    }

    #[allow(dead_code)]
    pub fn is_borrowed_list_valid(&self, cat: &Catalog) -> bool {
        for id in &self.borrowed_ids {
            if cat.get(&id).is_err() {
                return false;
            }
        }
        true
    }
}
